tests: boardfarm_plugins: port initial_ap_config

taurha opened the pull request:
Ported test_initial_ap_config from tests/test_flow.py for dockerized
boardfarm device and created prplMesh testsuite.

During porting:
* environment.py contents were used "as is", there's
  no strong necessity to modify it;
* boardfarm logs contains only output of commands issued by
  its devices, i.e. all actual test logs still generated in:
  <prplMesh-repo-root>/logs/*
  (for that reason, RadioDocker.wait_for_log still works)
* boardfarm logs can be found in directory "results" created
  in same dir bft was started from;

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

taurha commented:
That PR should be reviewed first: #1108

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,65 @@
+    def runTest(self):
+        '''
+        Main test logic must be implemented here.
+        '''
```
I think that such common comment should be replaced with description of test goals

MuteSpirit commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,65 @@
+        ]
+        failed_regex = []
+
+        wired_sniffer = sniffer.Sniffer(_get_bridge_interface(self.dev.board.unique_id))
```
Function `_get_gridge_interface` expected that docker network will have name `prplMesh-net-{unique_id}`, but in real test setup network may have another name template.
So i suggest 
1) set docker network name in test setup config
2) modify `_get_gridge_interface` to be able pass it docker network name instead of unique id

Commit:
tests: sniffer.py: replace global parameters usage

Usage of globals in Sniffer class makes it difficult for reusing.
To fix that, all globals were replaced by corresponding newly-added
class parameters.

Also, corresponding changes were performed in tests/environment.py,
to comply changes in Sniffer class.
(it's the only place it was using so far).

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
tests: _get_bridge_interface(): remove hardcoded network name

Changed function declaration to receive complete docker network name
instead of using unique_id as part of predefined one.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python3
```
This file doesn't have a main and isn't executable, so a shebang is useless.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+
+rootdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..'))
+
+logging.basicConfig(stream=sys.stdout, format='%(message)s')
```
I believe bft already does this so it shouldn't be needed, no?

taurha replied:
> I believe bft already does this so it shouldn't be needed, no?

Indeed, it's redundant snippet.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+class PrplMeshDocker(debian.DebianBox):
+    '''
+    Dockerized prplMesh device
+    '''
```
```suggestion
    '''Dockerized prplMesh device.'''
```

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+        self.args = args
+        self.kwargs = kwargs
+
+        config = kwargs["config"] if kwargs.get("config", None) else kwargs
```
```suggestion
        config = kwargs.get("config", kwargs)
```

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+    '''
+    model = ("prplmesh_docker")
+
+    def __init__(self, *args, **kwargs):
```
There is no call to the base class constructor anywhere, which seems weird to me...

I would actually expect to start with a call to the base class constructor, and then only handle the arguments which are specific to this class.

taurha replied:
> There is no call to the base class constructor anywhere, which seems weird to me...
> 
> I would actually expect to start with a call to the base class constructor, and then only handle the arguments which are specific to this class.

In our case, we using already existing boardfarm device as parent class, but when such parent class constructor sourced - we got bunch of exceptions from it.
It seems like existing boardfarm devices include additional functionality and requires some enhancement before we can use it as is - that is redundant for us.

So by inheriting existing boardfarm device, I just want to have access to parent methods that are absent in derived device class and mostly used by boardfarm internal logic.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+        self.cleanup_cmd = config.get("cleanup_cmd", None)
+        self.conn_cmd = config.get("conn_cmd", None)
+        self.mac = config.get("mac", self._get_rand_mac(oct_amount=4))
+        self.use_tcpdump = config.get("use_tcpdump", False)
```
Note that this option will go away (cfr. #1145) so don't bother adding it here. Just always assume true.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+        self.role = config.get("role", "agent")
+        self.cleanup_cmd = config.get("cleanup_cmd", None)
+        self.conn_cmd = config.get("conn_cmd", None)
+        self.mac = config.get("mac", self._get_rand_mac(oct_amount=4))
```
This will be removed (cfr. #1146) so don't bother adding it.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+                             stderr=subprocess.STDOUT, encoding="utf-8")
+        self.log_calls += cmd
+        self.log += res.stdout
+        logger.debug(res.stdout)
```
Avoid using shell for subprocesses, instead use proper individual command-line options as a list. We do need a shell for commands specified in the config file, but there shouldn't be a need for those in our specific PrplMeshDocker class.

I think this wrapper for subprocess is not terribly useful because it's not called very often and it just adds the logging bit.

Also, instead of subprocess, I think it's better to use bft_pexpect_helper because it handles timeouts.

taurha replied:
> Avoid using shell for subprocesses, instead use proper individual command-line options as a list. We do need a shell for commands specified in the config file, but there shouldn't be a need for those in our specific PrplMeshDocker class.
> 
> Also, instead of subprocess, I think it's better to use bft_pexpect_helper because it handles timeouts.

Applied in #1108.

I've already tried to use _bft_pexpect_helper_, but it's functionality seems to be in development and not works as expected. So instead I replaced _subprocess_ calls by _pexpect_ ones.

> I think this wrapper for subprocess is not terribly useful because it's not called very often and it just adds the logging bit.

IMO, without that wrapper the code will be inflated by similar snippets, so I would prefer to leave it to respect "Don't repeat yourself" principle.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+                                         "prplMesh-net-{}".format(self.unique_id))
+
+        if self.cleanup_cmd:
+            self._run_shell_cmd(self.cleanup_cmd)
```
Feels weird to call this in the constructor...

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+        docker_run_cmd = os.path.join(rootdir, "tools", "docker", "run.sh") \
+            + " --verbose --detach --force --name {}".format(self.name) \
+            + " --mac {}".format(self.mac) \
+            + " --network {}".format(self.docker_network)
```
This should be a tuple instead of a string.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+        if self.role == "both":
+            # Spawn dockerized controller + agent in single device
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
+            time.sleep(config.get("delay", 5))
```
Instead of doing config.get every time, add the delay as an attribute and use self.delay.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+
+        self.gw_iface = _get_bridge_interface(self.docker_network)
+
+        self.check_status()
```
It would be more appropriate to have this as a test instead of doing it in the constructor.

And that should be a separate commit.

taurha replied:
> It would be more appropriate to have this as a test instead of doing it in the constructor.
> 
> And that should be a separate commit.

That method is a "health check" and must be present in the device class:
it is used by boardfarm to indicate that spawned device instance is ready for test and also it is used after test to insure that device still operational.
So creating a testcase from that method will be wrong, as we have to run it after each test in boardfarm testsuite - that's a mess IMO.

arnout replied:
Actually, AFAICS in the source code of boardfarm, this method is only called after a failed test, as a way to distinguish between a failing test and a failing device.

But regardless, your point still stands: it doesn't make much sense to implement this as a separate test. And it *does* make sense to check the status in the constructor, since indeed boardfarm assumes that the device is fully operational when the constructor finishes.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+            # Spawn dockerized controller
+            self._run_shell_cmd(docker_run_cmd + " start-controller")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
```
As mentioned, controller-only won't be supported. Alternatively, we could support it but with `start-controller-agent` instead of `start-controller`.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:
```diff
@@ -0,0 +1,52 @@
+                "use_tcpdump": true,
+                "docker_network": "prplmesh-net",
+                "conn_cmd": "",
+                "radios": [
```
I would like to start out with a minimum config. It should have none of the options where we use the default (and conversely, the defaults should be what we need, e.g. delay).

The minimum config for now is just the top-level device without any subdevices.

Overall this needs to be split up into commits that add things in small steps. Now it's difficult to see why something is needed (and if it is really needed).

taurha replied:
> Overall this needs to be split up into commits that add things in small steps. Now it's difficult to see why something is needed (and if it is really needed).

Applied to #1108.
Added minimal config with empty subdevices list.
(`devices` list should be present even if it is empty)

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -0,0 +1,146 @@
+            # Spawn dockerized controller + agent in single device
+            self._run_shell_cmd(docker_run_cmd + " start-controller-agent")
+            time.sleep(config.get("delay", 5))
+            self.al_entity = ALEntityDocker(self.name, is_controller=True)
```
You should create two entities here: one for the agent and one for the controller.
```suggestion
            self.agent_entity = ALEntityDocker(self.name, is_controller=False)
            self.controller_entity = ALEntityDocker(self.name, is_controller=True)
```

For the other cases, we leave either agent_entity or controller_entity None.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,68 @@
+        wired_sniffer.start(self.__class__.__name__)
+
+        for radio in self.dev.board.al_entity.radios:
+            for regex in regex_list:
```
Although this looks like a nice refactoring at first sight, I don't really agree. I think it's harder to read this way than when it's explicitly separate checks, like is done in test_flows.

In general, try to take over the code from test_flows as literally as possible, and do any refactorings in a second stage.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -0,0 +1,68 @@
+                                                                entry[0],
+                                                                entry[1]
+                                                                )
+            raise Exception(msg)
```
Is raising an exception the only way to indicate failure?

taurha replied:
> Is raising an exception the only way to indicate failure?

Yes, `runTest()` methods doesn't have any return statement, so, according to boardfarm sources, the only way to indicate failure - throw exception.

arnout reviewed (commented):
Lots of comments but nothing major. Keep up the good work!

I probably missed some things in the PrplMeshDocker class so expect more comments in the second review cycle.

Commit:
boardfarm_prplmesh: prplmesh_docker.py: add wired sniffer

Configured wired sniffer for prplmesh_docker device in similar manner
it was made in test_flows.py.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
boardfarm_prplmesh: tests: port initial_ap_config to boardfarm

Ported initial_ap_config test from test_flows.py as boardfarm plugin.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
tests: utils.py: add class for common test functions

Added class which will store common functions used by boardfarm tests.
Copied test_flows.wait_for_log() to Utils class as Utils.check_log().

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

Commit:
boardfarm_prplmesh: prplmesh_config.json: add agent devices

Added basic description of two agent devices to .json config of
boardfarm prplMesh setup.

Signed-off-by: Oleksii Ponomarenko <o.ponomarenko@inango-systems.com>

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/prplmesh_config.json:
```diff
@@ -4,6 +4,17 @@
+        "devices": [
+            {
+                "name": "repeater1",
+                "type": "prplmesh_docker",
```
Actually, I believe the idea was to have repeater1 as DUT, and the controller as a peripheral device. So it would make more sense to structure it as
```
{
    "prplmesh_docker": {
        "name": "repeater1",
        "role": "agent",
        ...
       "devices": [
          {
            "name": "gateway",
            "role": "controller",
...
```

But this doesn't make any practical difference, so (nitpick).

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/devices/prplmesh_docker.py:
```diff
@@ -58,6 +60,8 @@ def __init__(self, *args, **kwargs):
             time.sleep(self.delay)
             self.agent_entity = ALEntityDocker(self.name, is_controller=False)
 
+        self.wired_sniffer = Sniffer(_get_bridge_interface(self.docker_network),
```
This is not correct - there should be a single, global, wired sniffer.

But then the problem, of course, is that you don't know the docker_network at the global scope.

So I think the solution is to only define the sniffer in the DUT specification. I'm not sure, however, how to distinguish the top-level device from the others... Maybe with an additional, explicit option in the JSON file? So in the JSON file we'd set `"wired_interface": "eth0"` to enable the sniffer on eth0. And we could have a special case `"wired_interface": "docker"` to use `self.docker_network`.

arnout commented on tests/boardfarm_plugins/boardfarm_prplmesh/tests/initial_ap_config.py:
```diff
@@ -4,13 +4,31 @@
-        """Main test logic must be implemented here."""
+        for dev in self.dev:
+            if dev.agent_entity:
+                dev.wired_sniffer.start(self.__class__.__name__)
```
This should be refactored to be done automatically before each test. But that can be done later.

arnout commented on tests/utils.py:
```diff
@@ -0,0 +1,20 @@
+    """Collection of common functions used by tests."""
+
+    @staticmethod
+    def check_log(entity_or_radio: Union[env.ALEntity, env.Radio], regex: str,
```
Personally, I much prefer the approach taken in current test_flows.py: define a base class that has these helper functions, and make the individual tests derive from that class. But I guess it's mostly a matter of taste. It does become important, however, if we ever need to keep any state - and I think we *will* need to do that.

arnout reviewed (approved):
All my comments are a little bit vague, so I'm OK with merging as-is and refactoring later. @oponomar please think about the idea of making a base class for the prplmesh tests and derive InitialApConfig from the base class, rather than introducing the Utils class.
